Documentation for Task Group Dynamic Dependencies (#1863)
uxlfoundation/oneTBB: doc/main/reference/task_group_dynamic_dependencies.rst

.. _task_group_dynamic_dependencies:

``task_group`` Dynamic Dependencies
===================================

.. note::
    To enable this extension, define the ``TBB_PREVIEW_TASK_GROUP_EXTENSIONS`` macro with a value of ``1``.

.. contents::
    :local:
    :depth: 2

Description
***********

The |full_name| implementation extends the
`tbb::task_group specification <https://oneapi-spec.uxlfoundation.org/specifications/oneapi/latest/elements/onetbb/source/task_scheduler/task_group/task_group_cls>`_

> with an API for defining predecessor-successor relationships between tasks
> meaning a successor task can begin execution only after all of its predecessors are completed.

Reviewer: ```suggestion
with an API for defining predecessor-successor relationships between tasks,
such that a successor task can begin execution only after all of its predecessors are completed.
```

> The Dynamic Dependencies API enables developers to define predecessor-successor relationships between tasks,
> meaning a successor task can begin execution only after all of its predecessors are completed.

Reviewer: ```suggestion
```

*Returns*: a reference to ``*this``.

.. code:: cpp

    task_completion_handle& operator=(const task_completion_handle& other);

Performs copy assignment from ``other`` to ``*this``. After the assignment, both refer to the same task.

*Returns*: a reference to ``*this``.

.. code:: cpp

    task_completion_handle& operator=(task_completion_handle&& other);

Performs move assignment from ``other`` to ``*this``. After the move, ``*this`` refers to the task previously referenced by ``other``, which is left empty.

*Returns*: a reference to ``*this``.

Observers
~~~~~~~~~

.. code:: cpp

    explicit operator bool() const noexcept;

*Returns*: ``true`` if ``*this`` references a task; otherwise, ``false``.

Comparison
~~~~~~~~~~

.. code:: cpp

    bool operator==(const task_completion_handle& lhs, const task_completion_handle& rhs) noexcept;

*Returns*: ``true`` if ``lhs`` and ``rhs`` reference the same task; otherwise, ``false``.

.. code:: cpp

    bool operator!=(const task_completion_handle& lhs, const task_completion_handle& rhs) noexcept;

Equivalent to ``!(lhs == rhs)``.

.. code:: cpp

    bool operator==(const task_completion_handle& t, std::nullptr_t) noexcept;

> *Returns*: ``true`` if ``t`` does not reference any task; otherwise, ``false``.

Author: Done

.. code:: cpp

    bool operator!=(const task_completion_handle& t, std::nullptr_t) noexcept;

Equivalent to ``!(t == nullptr)``.

.. code:: cpp

    bool operator==(std::nullptr_t, const task_completion_handle& t) noexcept;

Equivalent to ``t == nullptr``.

.. code:: cpp

    bool operator!=(std::nullptr_t, const task_completion_handle& t) noexcept;

Equivalent to ``!(t == nullptr)``.

Member Functions of ``task_handle`` Class
-----------------------------------------

.. code:: cpp

    ~task_handle();

Destroys the ``task_handle`` object and its associated task, if any.
If the associated task is involved in a predecessor-successor relationship, the behavior is undefined.

Member Functions of ``task_group`` Class
----------------------------------------

.. code:: cpp

    void run(task_handle&& handle);

> Schedules the task object associated with ``handle`` for execution, provided its dependencies are satisfied.

Reviewer: ```suggestion
Makes the task object associated with ``handle`` available for execution once its dependencies are satisfied. A task associated with a ``task_handle`` will never execute unless it is run.
```

    The ``run`` function does not wait for predecessors to complete.

.. code:: cpp

    static void set_task_order(task_handle& pred, task_handle& succ);

Registers the task associated with ``pred`` as a predecessor that must complete before the task associated with ``succ`` can begin execution.

It is thread-safe to concurrently add multiple predecessors to a single successor and to register the same predecessor with multiple successors.

It is thread-safe to concurrently add successors to both the task transferring its completion and the task receiving the completion.

The behavior is undefined in the following cases:

* Either ``pred`` or ``succ`` is an empty.
* The tasks referenced by ``pred`` and ``succ`` belong to different ``task_group`` instances.

.. code:: cpp

    static void set_task_order(task_completion_handle& pred, task_handle& succ);

Registers the task referenced by ``pred`` as a predecessor that must complete before the task associated with ``succ`` can begin execution.

It is thread-safe to concurrently add multiple predecessors to a single successor and to register the same predecessor with multiple successors.

> It is thread-safe to concurrently add successors to both the task transferring its completion and the task receiving the completion.

Reviewer: ```suggestion
It is thread-safe to concurrently add successors to both the task transferring its completion and the task receiving the completion.

It is thread-safe to concurrently add a predecessor while the ``task_handle`` associated with the same predecessor task is being run.
```

The behavior is undefined in the following cases:

* Either ``pred`` or ``succ`` is empty.
* The tasks referred by ``pred`` and ``succ`` belong to different ``task_group`` instances.

> * The task referred by ``pred`` was destroyed before begin submitted for execution.

Reviewer: ```suggestion
* The task referred by ``pred`` was destroyed before being submitted for execution.
```

*Returns*: ``true`` if ``t`` does not reference any task; otherwise, ``false``.

.. code:: cpp

    bool operator!=(const task_completion_handle& t, std::nullptr_t) noexcept;
    bool operator!=(std::nullptr_t, const task_completion_handle& t) noexcept;

Equivalent to ``!(t == nullptr)``.

Member Functions of ``task_handle`` Class
-----------------------------------------

.. code:: cpp

    ~task_handle();

Destroys the ``task_handle`` object and associated task if it exists.

.. admonition:: Extension

    If the associated task is involved in a predecessor-successor relationship, the behavior is undefined.

Member Functions of ``task_group`` Class
----------------------------------------

.. code:: cpp

    void run(task_handle&& handle);

Schedules the task object pointed by ``h`` for the execution.

.. admonition:: Extension

    If the task associated with ``h`` has predecessors, scheduling the task execution is postponed until all
    of the predecessors have completed, while the function returns immediately.

.. note::

    The failure to satisfy the following conditions leads to undefined behavior:
    * ``h`` is not empty.

>     * ``*this`` is the same ``task_group`` that ``h`` is created with.

Reviewer 2: ```suggestion
    The failure to satisfy the following conditions leads to undefined behavior:
    
    * ``h`` is not empty.
    * ``*this`` is the same ``task_group`` that ``h`` is created with.
```
For the proper rendering of the list.

Member Functions of ``task_arena`` Class
----------------------------------------

.. code:: cpp

    void enqueue(task_handle&& handle);

Enqueues a task owned by ``h`` into the ``task_arena`` for processing.

.. admonition:: Extension

    If the task associated with ``h`` has predecessors, scheduling the task execution is postponed until all
    of the predecessors have completed, while the function returns immediately.

The behavior of this function is identical to the generic version (``template<typename F> void task_arena::enqueue(F&& f)``),
except parameter type.

.. note::

    ``h`` should not be empty to avoid an undefined behavior.

Enqueues the task object associated with ``handle`` into ``task_arena`` for processing once its dependencies are satisfied.
A task associated with a ``task_handle`` will never execute unless it is run, enqueued or bypassed.

.. note::

    If the task associated with ``handle`` has incomplete predecessors, it will be scheduled for execution once all of them have completed.
    The ``enqueue`` function does not wait for predecessors to complete.

> 

Author: It was left from the old enqueue description. Removed
